lara/col: move misc utilities to col modules

lahm86 opened the pull request:
Resolves #3132.
Resolves #3134.

#### Checklist

- [x] I have read the [coding conventions](https://github.com/LostArtefacts/TRX/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] I have added a changelog entry about what my pull request accomplishes, or it is an internal change
- [x] I have added a readme entry about my new feature or OG bug fix, or it is a different change

#### Description

This re-homes the straggler col utility functions that were in `misc.c` into appropriate col modules, and as a result the bulk of them are now private. There are only two shared functions needed now altogether (ladder related); I moved wading control from `swim.c` to `land.c` because it shares much more in common with other land routines, and as a result the previous two shared functions from #3128 are now private again.

Apologies for the large diff, but it is mostly moving code around. The only major logical changes are for the two linked fixes: those came naturally with merging the related functions.

Main testing checks:
- The two linked bugs
- Swing cancel
- Ladders: climbing in all directions, getting on and off, to/from water and land
- Jumping forward to reach a ledge (succeeding, missing, plus combined with a ladder where Lara either reaches the top of the ledge or grabs the ladder below the top)
- A quick check of slope behaviour

lahm86 committed:
lara/col/climb: move Lara_TestClimbUpPos to module scope

lahm86 committed:
lara/col/climb: inline Lara_DoClimbLeftRight

lahm86 committed:
lara/col/climb: port Lara_HangTest to TRX

lahm86 committed:
lara/col/land: move Lara_TestWall to module scope

lahm86 committed:
lara/col/jump: port Lara_TestHangJump to TRX

lahm86 committed:
lara/col/jump: port Lara_TestHangJumpUp to TRX

lahm86 committed:
lara/col/jump: move Lara_TestEdgeCatch to module scope

lahm86 committed:
lara/col/jump: port Lara_SlideEdgeJump to TRX

lahm86 committed:
lara/col/jump: port Lara_DeflectEdgeJump to TRX

lahm86 committed:
lara/col/land: move wading to land module

Wading collision control has more in common with land routines,
so this moves it back to that module. This will allow further misc
functions to become private there.

lahm86 committed:
lara/col/land: inline Lara_SlideSlope

lahm86 committed:
lara/col/land: port Lara_TestSlide to TRX

lahm86 committed:
lara/col/land: port Lara_DeflectEdge to TRX

lahm86 committed:
lara/col/jump: port Lara_LandedBad to TRX

lahm86 committed:
lara/col/land: port Lara_HitCeiling to TRX

lahm86 committed:
lara/col/climb: port Lara_TestVault to TRX

lahm86 committed:
lara/col/climb: move Lara_TestClimbStance to module scope

lahm86 committed:
lara/col/climb: port ladder tests to TRX

This ports Lara_TestClimbPos and Lara_TestHangOnClimbWall to TRX
together as they are so intertwined, and performs some renaming and
stronger typing of test results.

lahm86 committed:
lara/col: move utilities to col module

rr- reviewed (commented):
A small request - can we isolate the fix commits into separate commits, without moving the functions around?

rr- commented on docs/tr2/README.md:
```diff
@@ -270,6 +270,7 @@ However, you can easily download them manually from these urls:
 - fixed Lara snapping to face forwards if she has a slight angle and action is pressed after using an airlock door
 - fixed Lara being able to equip guns and flares during in-game cutscenes
 - fixed the same boss item always being selected in Home Sweet Home, regardless of Lara's proximity
+- fixed Lara behaving erratically if she is killed while hanging from a ledge
```
Missing ticket number: 3134

lahm86 replied:
The ticket ref is in the changelog.
I'm having a lot of trouble trying to get the bug fix commits isolated (git conflicts). Would you prefer those in a separate PR and I'll just start this afresh? Or just add some comments?

lahm86 replied:
I was able to work around it. I changed the earlier commits to bake the bugs back in during porting, and there are now two final commits with the fixes isolated.

rr- replied:
Thanks! :heart:

If it's helpful, I use `git rebase -i` with `fixup` and `git revert` to manage commits in such scenarios effectively. Here's my process:

1. Edit the commit with the changes.
2. Reintroduce the bug temporarily and commit it (in a separate commit).
3. Revert that commit.
4. Continue rebasing, so the history looks like:
```
1234 move the function+fix the bug
2345 unfix the bug
3456 Revert of "unfix the bug"
517f more commits to follow
```
5. Rebase again to squash `unfix the bug` into the previous commit, updating the messages to focus on the move and fix.

I also compile the game in the steps inbetween, so that I know that each commit is actually buildable. (This is important for bisecting purposes.)
This technique prevents forward conflicts as from the perspective of the commits that follow the code hasn't changed at all. I can also try to move the fix commit where it makes more sense, but if this generates too many conflicts I let it be.

lahm86 replied:
Thanks, that looks great 😌 I attempted committing the fixes on a different branch then rebasing this branch on top of that but the conflicts wouldn't stop 😄 Will bear this mind next time (although I'm less in the habit these days of huge commits like this, it's completely my bad).

aredfan commented:
I've encountered a crash that affects TR1 only.

1. start up the game afresh
2. on the main menu, use `/play 2` or `/play 3`

lahm86 commented:
> I've encountered a crash that affects TR1 only.
> 
> 1. start up the game afresh
> 2. on the main menu, use `/play 2` or `/play 3`

I had this too and I think it's the ati save file (will investigate that separately as it's unrelated). Sorry, I should've mentioned. If you remove the save it should work.

lahm86 committed:
lara/col/jump: fix TR1 late grab position

This ensures to calculate Lara's bounds after having switching to her
grab animation, so that her Y position can be accurately determined.

Resolves #3132.

lahm86 committed:
lara/col/climb: fix perpetual swing cancel on death

This ensures Lara will let go of a ledge if she has died while on it,
rather than perpetually swing cancelling.

Resolves #3134.
